streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+    """
+    
+    MasterPlaylist.header = ''
+    with open(file_path, 'r') as master_playlist:
+      line = master_playlist.readline()
+      # Store each line in header until one of these tags is encountered.
+      while line and not line.startswith(('#EXT-X-MEDIA', '#EXT-X-STREAM-INF')):
+        # lstrip() will convert empty lines -> '' but will keep non-empty lines unchanged.
+        MasterPlaylist.header += line.lstrip()
+        line = master_playlist.readline()
+      else:
+        # Use this media playlist to also extract the MediaPlaylist header.
+        if line.startswith('#EXT-X-MEDIA'):
+          uri = _unquote(_extract_attributes(line)['URI'])
+        elif line.startswith('#EXT-X-STREAM-INF'):
+          uri = master_playlist.readline().strip()
+        else:
+          RuntimeError('No media playlist found in this master playlist')
+        master_playlist_dirname = os.path.dirname(file_path)
+        media_playlist_path = os.path.join(master_playlist_dirname, uri)
+        MediaPlaylist.extract_header(media_playlist_path)
+
+  @staticmethod
+  def concat_master_playlists(
+    master_playlists: List['MasterPlaylist']) -> 'MasterPlaylist':
+    
+    # SEMANTICS:
+      # xxx_playlist == one MediaPlaylist object
+      # xxx_playlists == multiple MediaPlaylist objects in the same master playlist.
+      # all_xxx_playlists == a list of (xxx_playlists).
+    all_txt_playlists: List[List['MediaPlaylist']] = []
+    all_aud_playlists: List[List['MediaPlaylist']] = []
+    all_var_playlists: List[List['MediaPlaylist']] = []
+    # The durations will be used to insert a gap in subtitles playlists when
+    # there is no applicable media to insert. Will also be used to calculate
+    # the average bandwidth.
+    durations: List[float] = []
+    
+    for master_playlist in master_playlists:
+      
+      txt_playlists: List['MediaPlaylist'] = []
+      aud_playlists: List['MediaPlaylist'] = []
+      var_playlists: List['MediaPlaylist'] = []
+      
+      for media_playlist in master_playlist.playlists:
+        stream_type = media_playlist.stream_info.get('TYPE', 'STREAM-INF')
+        if stream_type == 'SUBTITLES':
+          txt_playlists.append(media_playlist)
+        elif stream_type == 'AUDIO':
+          aud_playlists.append(media_playlist)
+        elif stream_type == 'STREAM-INF':
+          var_playlists.append(media_playlist)
+        else:
+          # TODO: We need a case for CLOSED-CAPTIONS(CC).
+          raise RuntimeError("TYPE={} is not regonized.".format(stream_type))  
+      
+      all_txt_playlists.append(txt_playlists)
+      all_aud_playlists.append(aud_playlists)
+      all_var_playlists.append(var_playlists)
+      durations.append(master_playlist.duration)
+    
+    master_hls = MasterPlaylist()
+    master_hls.playlists.extend(
+        MediaPlaylist.concat_sub(all_txt_playlists, durations))
+    
+    if all(not MediaPlaylist.inf_is_vid(inf_pl) for inf_pl in all_var_playlists):
```

Comment:
I'm having trouble reading this.  What's "inf" short for?  Variable names make a big difference in readability and maintainability.

Since "inf_pl" is an element from the list "all_var_playlists", could this fairly be called "var_playlist" instead?

Reply:
Oh, sorry about that, i just renamed all the `inf`s to `var`s today but overlooked this one.
I was calling a playlist that is a stream variant(`#EXT-X-STREAM-INF`) an `inf` playlist.